modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11y.tsx

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import './A11y.scss';
+
+import ClayIcon from '@clayui/icon';
+import ClayList from '@clayui/list';
+import ClayPopover from '@clayui/popover';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import ReactDOM from 'react-dom';
+
+import useA11y from '../../hooks/useA11y';
+
+import type {Violation as TViolation} from '../../hooks/useA11y';
+import type {A11yCheckerOptions} from './A11yChecker';
+
+const rectAttrs: Array<keyof DOMRect> = [
+	'bottom',
+	'height',
+	'left',
+	'right',
+	'top',
+	'width',
+];
+
+interface IRectState {
+	rect: DOMRect | undefined;
+	hasRectChanged: boolean;
+}
+
+const rectChanged = (a: DOMRect = {} as DOMRect, b: DOMRect = {} as DOMRect) =>
+	rectAttrs.some((prop) => a[prop] !== b[prop]);
+
+const useObserveRect = (
+	callback: (rect: DOMRect | undefined) => void,
+	node: Element | null
+) => {
+	const rafIdRef = useRef<number>();
+
+	const run = useCallback(
+		(node: Element, state: IRectState) => {
+			const newRect = node.getBoundingClientRect();
+
+			if (rectChanged(newRect, state.rect)) {
+				state.rect = newRect;
+
+				callback(state.rect);
+			}
+
+			rafIdRef.current = window.requestAnimationFrame(() =>
+				run(node, state)
+			);
+		},
+		[callback]
+	);
+
+	useEffect(() => {
+		if (node) {
+			run(node, {
+				hasRectChanged: false,
+				rect: undefined,
+			});
+
+			return () => {
+				if (rafIdRef.current) {
+					cancelAnimationFrame(rafIdRef.current);
+				}
+			};
+		}
+	}, [node, run]);
+};
+
+const Overlay = React.forwardRef<
+	HTMLDivElement,
+	React.ButtonHTMLAttributes<HTMLDivElement>
+>(({style, ...othersProps}, ref) =>
+	ReactDOM.createPortal(
+		<div {...othersProps} className="a11y-overlay" ref={ref} style={style}>
+			<div className="a11y-indicator">
+				<ClayIcon symbol="info-circle" />
+			</div>
+			<div className="a11y-backdrop" />
+		</div>,
+		document.body
+	)
+);
+
+type ViolationProps = {
+	modifyIndex: number;
+	target: string;
+	violations: Array<TViolation>;
+};
+
+function Violation({target, violations}: ViolationProps) {
+	const [visible, setVisible] = useState(false);
+	const [bounds, setBounds] = useState<React.CSSProperties>();
+
+	const node = document.querySelector(target);
```

Comment:
Can we wrap this in a useEffect for when the target changes? I'm not sure how often we expect this to render, but I think its good practice to not query the DOM in the main render thread of a component.

Reply:
In theory, we should keep it on a `useMemo` but I don't know if we will have an advantage here, when this component is rendered it is because the `modifyIndex` has changed which does not happen very often. Keeping on the hot path allows we to revalidate if the element still exists in the DOM as well, this is advantageous for this implementation, but we will hardly have many renderings on that component because if the target changes the instance of that component will be destroyed and a new one will be created because we use the target as `key`.

Reply:
If we only expect it to re-render when the `modifyIndex` changes, can we intentionally re-query when that prop changes? This way, by wrapping it with `useEffect(() => {}, [modifyIndex])`, we are intentionally communicating to anyone who reads this file in the future that the `node` is only expected to be changed when that prop changes. If we really do want to keep this reference for every render, another option would be to use it as a ref so that we can make sure we are referencing the same object/node between renders.

Also, relying on the `key` prop to destroy/create new components might be a little flakey and could run into problems in the future. Part of my thinking for this is, when a component's logic relies on a prop that is never used in the component file itself (such as `key`), it's hard to debug and figure out why a component has a bug or why it gets re-created vs re-used. Additionally, we might run into issues with future upgrades in React versions. If they  change how internal reconciliation works with key, whether it is more/less eager, we could run into issues. So, long story short, `key` is great but let's not rely on it too much since it's never explicitly used in the child component's file.

Reply:
> If we only expect it to re-render when the modifyIndex changes, can we intentionally re-query when that prop changes? This way, by wrapping it with useEffect(() => {}, [modifyIndex]), we are intentionally communicating to anyone who reads this file in the future that the node is only expected to be changed when that prop changes. If we really do want to keep this reference for every render, another option would be to use it as a ref so that we can make sure we are referencing the same object/node between renders.

Hmm, maybe we only need to do the query the first time since we have this behavior of the `key` on the target, so in theory we would always have the same `node` the problem with this is that maybe this `node` is no longer the same, for example: When you browse between the panels of the PageEditor Sidebar, we will lose the elements and if we go back we will have the same element again but with another node reference, in theory in our component we would have a reference to a node that no longer exists but we have the security that we do a revalidation before doing the aggregation of the violations where we remove the violations in which the target is not in the DOM, I think it can be safe to just use `const node = useMemo(() => document.querySelector(target), [])`, we can ignore the use of `modifyIndex` here as well because that means we only change the violations but the target remains the same.

> Also, relying on the key prop to destroy/create new components might be a little flakey and could run into problems in the future. Part of my thinking for this is, when a component's logic relies on a prop that is never used in the component file itself (such as key), it's hard to debug and figure out why a component has a bug or why it gets re-created vs re-used. Additionally, we might run into issues with future upgrades in React versions. If they change how internal reconciliation works with key, whether it is more/less eager, we could run into issues. So, long story short, key is great but let's not rely on it too much since it's never explicitly used in the child component's file.

Yeah but I think that here we don't have much to do, I think it's very difficult for this to change too soon but we use `key={target}` to help React find the element in the list, otherwise, he could suffer a lot more problems, sometimes the component is not being rendered or data is being lost, so `key` is valuable here because we constantly change the `Array` of violations and the same index can be different at different times.

I think we shouldn't be concerned in this case if this behavior will change in the future because if it does, it will be a big breaking change, and probably React will have ways to prevent it from happening because it is a common behavior for everyone when the list is being used and it is also React's own recommendation to use unique keys ids.

Reply:
> the problem with this is that maybe this node is no longer the same

Ah I see, I didn't realize that the node might not exist.

> const node = useMemo(() => document.querySelector(target), [])

More of a curiosity, but do you have any idea of the pro/con of using a ref instead? `useRef(document.querySelector(target));`


And yeah, I don't expect React to change anything major with `key`. My main point is that if changing the `key` causes a significant change in functionality, we might be relying on it too much. Meaning that if we change to `key={i}` and it causes a significant break, we have to be careful to communicate this to anyone who touches this file in the future. A future developer may be in the parent component moving logic around or adding a feature and not realize that the child component's logic with `key` relies heavily on the value that is passed in. My concern is more of a precaution for someone touching this file 3 years from now and less about how it works at the moment, so I may be overly-optimizing (or concerned) for nonexistent issues at the moment.

Reply:
> More of a curiosity, but do you have any idea of the pro/con of using a ref instead `useRef(document.querySelector(target));`

I think both are for specific use cases, for example:

- `useRef` will store the value only on the first rendering when it is declared, no matter how much your argument changes (in the case `document.querySelector(target)`), it also does not cause a new rendering and in this case, we will always be doing the query.
- `useMemo`, since it is a function as an argument, it will be stored and called only when the dependency reference changes during renderings, and when its value changes it causes a new rendering. It is usually recommended to use `useMemo` for something computationally expensive or when your component has many renderings.

> And yeah, I don't expect React to change anything major with key. My main point is that if changing the key causes a significant change in functionality, we might be relying on it too much [...]

Yeah, I think this is not an exaggeration but maybe I am giving the wrong impression that I am depending on `key` to rebuild the instance and not have problems, this is not quite the case, I was referring to having the` target` as a dependency will not have any effect, because when the `target` changes React will not take advantage of the instance, it will destroy and create a new one. So when we have the `useMemo(() => {}, [target])` it's the same as `useMemo` having the zero deps. Does it make sense for you?

Reply:
> useMemo, since it is a function as an argument, it will be stored and called only when the dependency reference changes during renderings

Yeah that makes sense, but for our case since we are passing `[]` as out dependency list we are saying to only store the value once (at mount) and so that made me think using a ref in this case would make no difference since both would reference the same object in memory.

But now that I think about it more, it is slightly more declarative to have `useMemo` with an empty dependency list since ref's are mutable. The mental model of each would be...
`const A = useMemo(() => someVal, [])`: The value of `A` never changes
`const B = useRef(someVal);`: The value of `B` might change

--------------

> So when we have the useMemo(() => {}, [target]) it's the same as useMemo having the zero deps

Yeah this makes sense to me, my line of thinking here was that by using `const node = useMemo(() => document.querySelector(target), [target])` instead of `key={target}`, we are making sure it's a bit more declarative. Because right now, _and correct me if I am wrong here_, if we wanted to use the `<Violation />` component in another file, we would have to make sure and use `key={target}` each time we use it so that it re-creates when `target` changes.

Reply:
I could also entirely be reading wrong here, my mind does not feel extra sharp today, this is how fast I am moving ![](https://media.giphy.com/media/3NtY188QaxDdC/giphy.gif)

Reply:
> Yeah that makes sense, but for our case since we are passing [] as out dependency list we are saying to only store the value once (at mount) and so that made me think using a ref in this case would make no difference since both would reference the same object in memory.

I don't know if I understand correctly but in the case of `useRef` is that even though the reference is the same compared to having another `useMemo` that makes the same query, in `useRef` we will always be doing the query for the selector in the DOM the return of the node can be the same but the computational cost will still exist.

> `const B = useRef(someVal);`: The value of B might change

I would not say that the value of B might change because although `someVal` can change, `useRef` will not update, only if we redefine its value explicitly `B.current = someVal`.

A good comparison of `useRef` is as we did on components with class, adding a new property to the instance of the class but which will have no effect during the component's life cycle but the value will be persisted until disassembly.

> Yeah this makes sense to me, my line of thinking here was that by using const node = useMemo(() => document.querySelector(target), [target]) instead of key={target}, we are making sure it's a bit more declarative. Because right now, and correct me if I am wrong here, if we wanted to use the <Violation /> component in another file, we would have to make sure and use key={target} each time we use it so that it re-creates when target changes.

Hmm yes I think this is a little right your mentality about it. If we want to avoid making queries in the DOM every time a rendering occurs using `useMemo` without dependency we will have this effect when` <Violation /> `is used outside, as it is today we would not have the problem but we have the computational cost of searches the DOM every time a rendering happens, so, in that case, I agree with you that having `const node = useMemo(() => document.querySelector(target), [target])` would be much safer and we get the middle ground.

Reply:
> in useRef we will always be doing the query for the selector in the DOM the return of the node can be the same but the computational cost will still exist.

Yeah the computation of querying the DOM will always happen for useRef on the initial initialization of the component. And for the case of useMemo, it will query the DOM only on mount and when the dependencies change. So in the case of useRef vs useMemo with no dependencies, there is only a slight difference in how they work. [This was sort of the question I was thinking through](https://stackoverflow.com/questions/58810036/empty-dependencies-with-usememo-or-usecallback-vs-useref), more theoretical than it was for your specific code.


> const B = useRef(someVal);: The value of B might change

What I meant by this is that when you use a ref, it is possible that the value is mutated in the logic flow of your code and you can't know if it's mutated or not unless you read the entire flow of logic. By using a memo with no dependencies, we can be sure it will never mutate and only compute on mount. Again, this is just me thinking through the mental model of the two and how I would read through some react code.

--------

> so, in that case, I agree with you that having const node = useMemo(() => document.querySelector(target), [target]) would be much safer and we get the middle ground.

👍  👍  Sounds good to me and if we end up having issues or realize we need to do something else in the future we can always change it.

Reply:
> What I meant by this is that when you use a ref, it is possible that the value is mutated in the logic flow of your code and you can't know if it's mutated or not unless you read the entire flow of logic. By using a memo with no dependencies, we can be sure it will never mutate and only compute on mount. Again, this is just me thinking through the mental model of the two and how I would read through some react code.

Oh yeah, makes sense